# HR Analytics DATA 
Sourced From [Kaggle Dataset- Employee Attrition and Factors](https://www.kaggle.com/datasets/thedevastator/employee-attrition-and-factors)

In [57]:
#Senbagaraman - https://github.com/senbagaraman04/ml-basics/tree/main/2023

### Import Libraries

In [58]:
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt


### Read the Data

In [59]:
df = pd.read_csv('./HR_Analytics.csv')
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [60]:
df.shape

(1470, 35)

 1470 rows with 35 columns 

In [61]:
#List all the columns in the file 
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

### Find the Columns with numbers and Strings 

In [62]:

# def findCOlumnTypes():
#     int_columns = []
#     object_columns = []
#     for col in df.columns:
#         if (df[col].dtype ==  np.int64):
#              int_columns.append(df[col].name)
#         else:
#             object_columns.append(df[col].name)

#     print(int_columns)
#     print("***************")
#     print(object_columns)


In [63]:
df.dtypes
#findCOlumnTypes()

cat_coulmns  = df.select_dtypes(['object']).columns

num_coulmns  = df.select_dtypes(['int64']).columns

print(cat_coulmns)
print(num_coulmns)


Index(['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender',
       'JobRole', 'MaritalStatus', 'Over18', 'OverTime'],
      dtype='object')
Index(['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome',
       'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')


### Find all the values present in the categorical columns

In [64]:
for col in cat_coulmns:
    print(df[col].value_counts())
    print("**************")

No     1233
Yes     237
Name: Attrition, dtype: int64
**************
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64
**************
Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64
**************
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64
**************
Male      882
Female    588
Name: Gender, dtype: int64
**************
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: JobRole, dtype: int64
**************
Married     673
Single      470
Divorced    327
Name: MaritalSt

Table compile from running the above code ```df[cat_coulmns].value_counts()```

| Column Name | Values and Count |
|-----|----|
| Attrition | No: 1233 Yes: 237 |
| BusinessTravel | Travel_Rarely  1043::Travel_Frequently 277::Non-Travel            150|
| Department |R&D 961::Sales 446::HR 63|
| EducationField | Life Sciences  606::Medical 464::Marketing  159::Technical Degree 132::Other 82::HR 27 |
| Gender | M 882::F 588|
| JobRole | Sales Executive 326::Research Scientist  292::Laboratory Technician  259::Manufacturing Director  145::Healthcare Representative    131::Manager 102::Sales 83::Research Director  80::HR 52 |
|MaritalStatus| M 673:: S 470::D 327|
|OverTime|No  1054::Yes 416|


### Encoding

In [65]:
# Attrition Column: 
#df[cat_coulmns] = df[cat_coulmns].apply(lambda x: pd.factorize(x)[0])



Attrition  BusinessTravel  Department              EducationField    Gender  JobRole                MaritalStatus  Over18  OverTime
No         Travel_Rarely   Research & Development  Life Sciences     Male    Research Scientist     Married        Y       No          19
                                                   Medical           Male    Laboratory Technician  Married        Y       No          17
                           Sales                   Marketing         Male    Sales Executive        Married        Y       No          15
                           Research & Development  Life Sciences     Male    Laboratory Technician  Married        Y       No          15
                                                                             Research Scientist     Single         Y       No          14
                                                                                                                                       ..
                           Human Resourc